### DISCUSSION TITLE
Cannot get gcno sloppiness to work

### DISCUSSION THREAD

**Discussion by BenPortner** [ccache/ccache#1444] [Support]
## How to reproduce

Create a dummy .cpp file:

```cpp
// test_ccache.cpp
#include <iostream>

int main() {
    std::cout << "Testing ccache..." << std::endl;
    return 0;
}
```

Compile with c++:

```bash
rm -rf build
mkdir build
export CCACHE_BASEDIR=$(pwd)
export CCACHE_SLOPPINESS=gcno_cwd
cd build
ccache c++ --coverage -c ../test_ccache.cpp
```

Test if gcno contains absolute base path:

```bash
# execute in $CCACHE_BASEDIR/build
grep -r $CCACHE_BASEDIR .
```

## Actual behavior
Grep returns hits in the .o and .gcno file:

```
grep: ./test_ccache.o: binary file matches
grep: ./test_ccache.gcno: binary file matches
```

These hits cause ccache misses when compiling in other directories (see https://github.com/ccache/ccache/issues/1032) 

## Expected behavior

Grep returns no hits. Ccache reuses artefacts when compiling in other directories.

## Environment
I tried with ccache versions 4.6.1 and 4.9.1
c++ 11.4.0
Ubuntu 22.04

## Other info

Grep returns no hits when omitting the `--coverage` flag while compiling. 

Debug log: [test_ccache.o.20240429_154450_047281.ccache-log.txt](https://github.com/ccache/ccache/files/15150045/test_ccache.o.20240429_154450_047281.ccache-log.txt)

**Comment by jrosdahl** [comment]
@BenPortner: gcno_cwd sloppiness is about `.gcno` files, which are generated when compiling with `-ftest-coverage`. The `.gcno` file embeds CWD by default, which ccache can be sloppy about.

When compiling with `-fprofile-arcs` the object file will include an absolute path to the `.gcda` file that will be created when executing the program later. (The `.gcda` path is the absolute path to the object file with `.o` replaced with `.gcda`.) Ccache cannot be sloppy about that.

`--coverage` is a synonym for `-fprofile-arcs  -ftest-coverage`, so results cannot be reused between different build directories due to use of `-fprofile-arcs`.

Perhaps the manual could be more clear about this. Any suggestions?

**Comment by BenPortner** [comment]
Hello @jrosdahl,

thank you for the detailed explanation. This is probably a stupid question but: If the path to the .gcda file is the path to the .o file with the file extension replaced...then why is the absolute path needed? They are in next to each other in the same directory, right?

Concerning the manual, here is a first shot (changes highlighted in cursive):

> By default, ccache will include the current working directory in the hash when producing a .gcno file _(when compiling with -ftest-coverage)_. This is because GCC 9+ includes the current working directory in the .gcno file. The gcno_cwd sloppiness makes ccache not hash the current working directory so that you can get cache hits when compiling in different directories, with the tradeoff of potentially getting an incorrect directory in the .gcno file. gcno_cwd also disables hashing of the current working directory if -fprofile-abs-path is used. _Note that sloppiness will not work when using the --coverage flag, as discussed [here](https://github.com/ccache/ccache/discussions/1444)_.

**Reply by jrosdahl** [reply]
> why is the absolute path needed

The `.gcda` path is embedded in the object file by the compiler to know where to write the `.gcda` file when the program is executed. That is, if you move the program to another directory, change to that directory and run the program there then the `.gcda` file will be written next to the  original object file location, not the CWD.

> here is a first shot

Thanks. I committed a variant of it in 350d74cce91b6156b5ad4b83847cd06f79176148.

**Reply by BenPortner** [reply]
Thanks, it's perfectly clear now.

Sooo is there any way we could get a gcda sloppiness for people who don't plan to move their binaries around (at least not before generating the coverage reports)?

**Reply by jrosdahl** [reply]
Moving the binary was maybe not the ideal example. :slightly_smiling_face:  The problem is that is if you compile an object file in a directory and then use it in a build in another directory (which is what would happen if ccache allowed it) then the resulting program will write the `.gcda` file to the wrong directory.

That is, if ccache would allow "gcda sloppiness" something like this would happen:

```console
$ echo 'int main(){}' >example.c
$ cd dir1
$ ccache gcc -fprofile-arcs -c ../example.c
$ strings example.o | grep dir
/path/to/dir1/example.gcda
$ cd ../dir2
$ ccache gcc -fprofile-arcs -c ../example.c
$ strings example.o | grep dir
/path/to/dir1/example.gcda
$ gcc example.o -lgcov -o example
$ ./example
$ ls example.gcda
ls: cannot access 'example.gcda': No such file or directory
$ ls ../dir1/example.gcda
../dir1/example.gcda
```

It would be easy to add a gcda sloppiness, but I don't see how that would be useful to anyone when the program will be broken? (If you don't care about profiling information, compile without `-fprofile-arcs`?)

_Actually_, the above already happens for relative paths, which is a bug... Fixed in 350d354b4502ddfc2638362be9f0d48862659900.

**Reply by BenPortner** [reply]
Thanks again for the helpful example! I understand the problem better now. However, it looks to me like the problem occurs because `example.o` still contains the absolute path (`/path/to/dir1/example.gcda`). If it contained the relative path (`./example.gcda`) then the gcda file would always be written next to the correct example.o file (dir1/example.o in the first case, dir2/example.o in the second case), am I wrong? At least that's how I would want the gcda-sloppiness to work.

**Reply by jrosdahl** [reply]
> If it contained the relative path (./example.gcda) then the gcda file would always be written next to the correct example.o file

That's true. However, I'm not aware of any way to tell the compiler to insert a relative path.

**Reply by BenPortner** [reply]
I played around with some of the gcc options but I couldn't find a way to achieve this either. Instead of digging further to come up with some (most likely pretty awful) hack it is easier to work from the other end and ensure that compilation happens in the same location always. This can be achieved by mounting the source code into a container at a pre-defined location. Thanks for your great patience as always @jrosdahl. You may close the discussion if you like.
